mcs/class/corlib/System/TimeZoneInfo.Unity.cs

```diff
@@ -45,90 +22,100 @@ namespace System {
 	public partial class TimeZoneInfo {
 		enum TimeZoneData
 		{
-		    DaylightSavingStartIdx,
-		    DaylightSavingEndIdx,
-		    UtcOffsetIdx,
-		    AdditionalDaylightOffsetIdx
+			DaylightSavingStartIdx,
+			DaylightSavingEndIdx,
+			UtcOffsetIdx,
+			AdditionalDaylightOffsetIdx
 		};
 		
 		enum TimeZoneNames
 		{
-		    StandardNameIdx,
-		    DaylightNameIdx
+			StandardNameIdx,
+			DaylightNameIdx
 		};
+
+		static AdjustmentRule CreateAdjustmentRule(int year, out Int64[] data, out string[] names, string standardNameCurrentYear, string daylightNameCurrentYear)
+		{
+			if(!System.CurrentSystemTimeZone.GetTimeZoneData(year, out data, out names))
+				return null;
+			DateTime startTime = new DateTime (data[(int)TimeZoneData.DaylightSavingStartIdx]);
+			DateTime endTime = new DateTime (data[(int)TimeZoneData.DaylightSavingEndIdx]);
+			TimeSpan daylightOffset = new TimeSpan (data[(int)TimeZoneData.AdditionalDaylightOffsetIdx]);
+
+			/* C# TimeZoneInfo does not support timezones the same way as unix. In unix, timezone files are specified by region such as
+			 * America/New_York or Asia/Singapore. If a region like Asia/Singapore changes it's timezone from +0730 to +08, the UTC offset
+			 * has changed, but there is no support in the C# code to transition to this new UTC offset except for the case of daylight
+			 * savings time. As such we'll only generate timezone rules for a region at the times associated with the timezone of the current year.
+			 */
+			if(standardNameCurrentYear != names[(int)TimeZoneNames.StandardNameIdx])
+				return null;
+			if(daylightNameCurrentYear != names[(int)TimeZoneNames.DaylightNameIdx])
+				return null;
+
+			TransitionTime dlsTransitionStart = TransitionTime.CreateFixedDateRule(new DateTime(1,1,1).Add(startTime.TimeOfDay),
+																						startTime.Month, startTime.Day);
+			TransitionTime dlsTransitionEnd = TransitionTime.CreateFixedDateRule(new DateTime(1,1,1).Add(endTime.TimeOfDay),
+																						endTime.Month, endTime.Day);
+
+			AdjustmentRule rule = TimeZoneInfo.AdjustmentRule.CreateAdjustmentRule(new DateTime(year, 1, 1),
+																				new DateTime(year, 12, DateTime.DaysInMonth(year, 12)),
+																				daylightOffset,
+																				dlsTransitionStart,
+																				dlsTransitionEnd);
+			return rule;
+		}
+
 		static TimeZoneInfo CreateLocalUnity ()
 		{
 			Int64[] data;
-		    string[] names;
-		    if (!System.CurrentSystemTimeZone.GetTimeZoneData (1973, out data, out names))
+			string[] names;
+			int currentYear = DateTime.UtcNow.Year;
```

Comment:
No chance of infinite recursion here, right?

Reply:
No, looks like DateTime.UtcNow calls into an icall to get the current system time..converts that to ticks and passes it into the DateTime(<ticks>) constructor...should be no change of recursion, and I'm not seeing it myself.